New inverse stream map to accelerate convergence

unalmis opened the pull request:
![PR Age](https://img.shields.io/github/issues/detail/age/PlasmaControl/DESC/1919?label=Ready%20for%20merge)

## Inverse stream maps
- [x] Significantly improves convergence of inverse stream maps. Resolves #1922 to my satisfaction.
- [x] Truncation rule to remove aliasing error at the shortest wavelengths where the signal to noise ratio is lowest.
- [x] Resolves #1928 and increases default objective resolution for `Y_B`.
- [x] Method to plot frequency spectrum of inverse stream map.

## Improvements
- [x] Check-pointing to increase speed and reduce memory consumption of reverse mode differentiation #1347.
- [x] Adds ``low_ram`` mode which is same speed and less memory for `objective.compute`, but slower for `objective.grad` since JAX is poor at iterative algorithms.
- [x] Fully resolves #1864 by avoiding materialization of a large tensor in memory. Previously, we had closed the issue by adding nuffts as a workaround. The improvement here actually solves the JAX regression.
- [x] Reuses some computations in identifying bounce points to make more efficient.
- [x] Increase cache hits, fusing, and reduce floating point error in computing bounce points (very important for accurate integrals).
- [x] Transforms an improper field line integral to one on a compact domain where the integrand is periodic to achieve faster convergence.
- [x] Improves performance complexity of `interp_to_argmin` for `Bounce2D` from fourth order to spectral as required for Alpert quadrature.
- [x] Resolves #1389.

## Usability
- [x] Updates plotting utilities to accept `kwargs` as was needed for my article.
- [x] Method to compute integrals over plasma volume and velocity space in batches is now added to the public API. People kept asking how to use this private function and also used it for their work, so better to make it public.
- [x] Method to make adding new bounce integral objectives simpler. Private for developers only.
- [x] Unifies default parameter setup to resolve #1898 for bounce integrals.
- [x] Resolves #1388.

## Bugs
- [x] Fixes bug in ``test_compute_everything``.
- [x] Fixes inverse stream map convergence.


## Benchmarks
Just go to #2026 and run `effective_ripple_profile.py`. You will see the large performance improvement from `master`. The CI benchmarks do not reveal this because those benchmarks are essentially just noise. Note that, using the same parameter inputs, the resolution of this branch is also higher than `master` due to the faster convergence.

- If you set `use_bounce1d=True` on that script, you will run out of memory as expected since it is an inferior approach (as expected, you get the OOM in the jacobian before you compute a single bounce integral).
- If you set ``nufft_eps=0``, you need 175 GB to run that script on `master` (you'll get an OOM and JAX will tell you it needs 175GB), but only 34 GB on this branch.
- Using nuffts, the script requires only 6.5 GB on this branch. 

# Examples

## HELIOTRON

### `Master` branch
<img width="480" height="360" alt="test_theta_chebyshev_HELIOTRON" src="https://github.com/user-attachments/assets/9ae1e6c5-983f-48a2-a805-b28fb1aec1e1" />

### This branch
<img width="480" height="360" alt="test_delta_fourier_chebyshev_HELIOTRON" src="https://github.com/user-attachments/assets/3773b5bb-6230-4362-8136-dd91543ac217" />

## W7-X

### `Master` branch
<img width="480" height="360" alt="test_theta_chebyshev_W7-X" src="https://github.com/user-attachments/assets/f2ba109c-0c72-4efc-b432-ed2993ac0795" />

### This branch
<img width="480" height="360" alt="test_delta_fourier_chebyshev_W7-X" src="https://github.com/user-attachments/assets/ef9a642a-7bbc-41b8-a615-7bde6849856d" />

### NCSX

### `Master` branch
<img width="480" height="360" alt="test_theta_chebyshev_NCSX" src="https://github.com/user-attachments/assets/b3d4371d-0539-4c4b-a7a8-0e882258f2f7" />

### This branch
<img width="480" height="360" alt="test_delta_fourier_chebyshev_NCSX" src="https://github.com/user-attachments/assets/5c92c2c2-22de-47d9-bfd7-f3fbed40bb43" />

## Removal of spectral aliasing

<img width="480" height="360" alt="Figure_1" src="https://github.com/user-attachments/assets/c4e78659-b03d-41cb-8e0f-809066b67989" />

# Dynamic shapes

- [x] resolves #1303 .
- [x] Activates newton step to find bounce points.
  - It can be shown there exists O(sqrt(epsilon)) error for bounce integrals with 1/v_ll where epsilon is error of bounce point. For v_|| integrals error is conveniently O(epsilon times sqrt(epsilon)). Hence, the `spline` method would require thousands of knots per transit for just a couple digits of accuracy, and it would stop convergence at epsilon<=1e-5 (so sqrt(epsilon) <=3 digits) error due to condition number. Of course, the `spline=False` method has always computed the points with spectral accuracy and has very fast convergence after #1919 ; that method converges to epsilon = machine precision without the newton step. With the newton step, fast convergence is achieved with the spline method as well.
  - I suspect fast ion confinement optimization will be easier now.

## Benchmarks
Here is a timing benchmark on my CPU with `nufft_eps=1e-6`. Prior to this PR, every adjoint call to nufft1 took `>= 1 second` and the full computation was `34 seconds`. Now every adjoint call to nufft1 is `250 milliseconds`, and the full computation is `14 seconds`. These `improvements become larger` as the `sparsity` grows and error tolerance parameter for the nuffts epsilon tends to 0. Likewise, the _improvement_ grows linearly with the `problem size`. As this is called within an optimization loop where time and memory are tight, the improvement is significant.

### Before
<img width="510" height="135" alt="Screenshot From 2026-03-29 15-22-56" src="https://github.com/user-attachments/assets/0baf9a88-b775-4a9c-bb39-db568099aae7" />
<img width="510" height="135" alt="Screenshot From 2026-03-29 15-22-36" src="https://github.com/user-attachments/assets/2bc1f730-56af-43ad-9c5e-45e9991c0bd1" />

### After
<img width="510" height="135" alt="Screenshot From 2026-03-29 15-12-35" src="https://github.com/user-attachments/assets/306a6c34-32a3-46e3-8b8c-c4f259a32169" />
<img width="510" height="135" alt="Screenshot From 2026-03-29 15-15-59" src="https://github.com/user-attachments/assets/db12ecb3-210a-4732-a5fc-666ba47475d8" />

unalmis commented on desc/compute/_fast_ion.py:
```diff
@@ -168,7 +170,7 @@ def Gamma_c(data):
         def fun(pitch_inv):
             points = bounce.points(pitch_inv, num_well)
-            v_tau, drift1, drift2 = bounce.integrate(
+            vτ, drift1, drift2 = bounce.integrate(
```
None of these variables are callable by users or developers. So no one can ever type them anyway. In the bounce integral source code at least, using unicode has decreased my development time because I can confirm correctness much easier.

dpanici replied:
I'm with Rory on this, no unicode variables

unalmis replied:
and the wistell code uses unicode fwi if interested

unalmis replied:
https://gitlab.com/wistell/StellaratorOptimizationMetrics.jl/-/blob/master/src/GammaC/Functions.jl?ref_type=heads

unalmis commented on desc/batching.py:
```diff
@@ -272,9 +272,6 @@ def vmap_chunked(
 
     """
     in_axes, argnums = _parse_in_axes(in_axes)
-    if isinstance(argnums, int):
```
duplicate code

YigitElma commented on desc/compute/_neoclassical.py:
```diff
@@ -89,14 +88,15 @@
         Precomputed transform matrices "dct spline", "dct cfl", "dft cfl".
         This parameter is intended to be used by objectives only.
         """,
+    "theta": "",
```
This is deprecated, right?

unalmis replied:
yes

YigitElma commented on desc/equilibrium/coords.py:
```diff
@@ -453,43 +451,57 @@ def _partial_sum(lmbda, L_lmn, omega, W_lmn, iota):
         Spectral coefficients for λ.
     lmbda : Transform
-        Transform for λ built on DESC coordinates [ρ, θ, ζ].
+        Transform for λ built on DESC coordinates (ρ, θ, ζ) uniformly spaced in θ.
```
Just to clarify, the coordinate system doesn't matter here, right? You create a transform object with some `Grid` with uniform spacing in the poloidal direction, depending on the inbasis outbasis values whatever poloidal angle this transform correcponds to will change. Just noting this since `built on DESC coordinates (ρ, θ, ζ) uniformly spaced in θ.` might create confusion (say if you have a different inbasis)

unalmis replied:
The docstring is correct. You need rho, theta, zeta regardelss of inbasis.

YigitElma commented on desc/integrals/_interp_utils.py:
```diff
@@ -66,9 +66,9 @@ def cheb_pts(n, domain=(-1, 1), lobatto=False):
 
-def fourier_pts(n):
+def fourier_pts(n, domain=(0, 2 * jnp.pi)):
     """Get ``n`` Fourier points in [0, 2π)."""
```
```suggestion
    """Get ``n`` Fourier points in domain."""
```
?

YigitElma commented on desc/integrals/basis.py:
```diff
@@ -256,14 +251,167 @@ def compute_cheb(self, x):
+        Domain for y coordinates. Default is [-1, 1].
+    lobatto : bool
+        Whether ``f`` was sampled on the Gauss-Lobatto (extrema-plus-endpoint)
+        instead of the interior roots grid for Chebyshev points.
```
In addition to the error below, the docstring can also mention that jax doesn't have dct with lobatto points.

unalmis commented on desc/integrals/_bounce_utils.py:
```diff
@@ -467,145 +463,170 @@ def argmin(z1, z2, f, ext, g_ext):
+        periodic in θ. Hence, θ mod 2π is simple to recover from the interpolation of
+        δ on a single branch (α, ζ) ∈ [0, 2π)².
+
+    The field line label α changes discontinuously along a magnetic field line.
```
it is discontinuous across any branch cut. Toroidal or poloidal

unalmis commented on desc/equilibrium/coords.py:
```diff
@@ -500,40 +512,70 @@ def _map_clebsch_coordinates(
-        return target_k - target
+    # Root finding for θₖ such that θₖ + (λ−ιω)(ρ,θₖ,ζ) - ε = 0.
+    def rootfun(θ, varepsilon, q_m):
+        q = (jnp.exp(1j * modes * θ) * q_m).real.sum()
```
i can change this variable here, but I am not changing the bounce integral code

unalmis commented on desc/integrals/basis.py:
```diff
@@ -256,14 +251,167 @@ def compute_cheb(self, x):
         return cheb
 
 
+class ChebyshevSeries(IOAble):
```
it's not public and all this interpolation stuff will be moved to a new library in the future #1388

f0uriest commented on desc/equilibrium/coords.py:
```diff
@@ -453,43 +448,57 @@ def _partial_sum(lmbda, L_lmn, omega, W_lmn, iota):
+    varepsilon : jnp.ndarray
+        Shape should broadcast with shape (num ρ, num poloidal, num ζ).
+        Optional meshgrid of values ε = α + ι ζ.
+        This meshgrid need not be a tensor-product grid.
```
what exactly does this mean? in most places we've been using meshgrid to mean a tensor product grid

unalmis replied:
Meshgrids are more general than tensor-product grids. For example, if you iterate along the the second axis of the quantity ``varepsilon.reshape(num rho, num poloidal, num zeta)``, then the values of ``varepsilon[:, i, :]`` can be different for each ``i``. In contrast, on a tensor-product grid, varepsilon has to be the same for each `i`.

The purpose of the comment is basically to inform the user that we support tensor product grids in (rho, alpha, iota zeta) in addition to tensor product grids in (rho, alpha, zeta). 

If you want to perform a coordinate mapping for a single poloidal transit on every flux surface, then giving ``varepsilon`` as a meshgrid in (rho, alpha, zeta) works but it can't be a tensor-product grid in (rho, alpha, zeta) unless iota is constant wrt to rho.

f0uriest commented on desc/equilibrium/equilibrium.py:
```diff
@@ -1134,7 +1134,13 @@ def need_src(name):
-                key: grid1dr.copy_data_from_other(data[key], grid, surface_label="rho")
+                key: (
+                    data[key]
+                    if (key == "V_r/psi_r")
```
why can't this be copied like the others?

f0uriest commented on desc/objectives/_fast_ion.py:
```diff
@@ -91,22 +95,22 @@ class GammaC(_Objective):
     num_well : int
         Maximum number of wells to detect for each pitch and field line.
-        Giving ``None`` will detect all wells but due to current limitations in
+        Giving ``-1`` will detect all wells but due to current limitations in
```
is the default of `None` the same as `-1` or does it use the `A+iota*C` heuristic?

unalmis replied:
In this PR, I have changed the default `None` to be very safe heuristic that we set. If the user has unlimited compute, and does not trust our heurstic, they can set it to `-1` to ensure all wells are found.

f0uriest commented on desc/objectives/_neoclassical.py:
```diff
@@ -67,17 +67,18 @@ class EffectiveRipple(_Objective):
         Default grid samples the boundary surface at ρ=1.
     X : int
-        Poloidal Fourier grid resolution to interpolate the poloidal coordinate.
+        Poloidal Fourier grid resolution to interpolate the angle.
```
"angle" here means delta = theta - alpha?

unalmis replied:
Yes

f0uriest commented on desc/equilibrium/coords.py:
```diff
@@ -453,43 +448,57 @@ def _partial_sum(lmbda, L_lmn, omega, W_lmn, iota):
     lmbda,
+    varepsilon=None,
+    inbasis=("rho", "alpha", "zeta"),
+    outbasis=("rho", "theta", "zeta"),
```
im a bit confused by inbasis and outbasis here. isn't the input radial coordinate always iota and input toroidal coordinate always zeta? and the output is always only the poloidal angle? I think it would be clearer if inbasis and outbasis were just single coordinates, not triplets (basically to label what `poloidal` is and what the output should be)

unalmis replied:
I was trying to match the eq.map_coordinates api. i can change as you have requested to just one variable.

f0uriest commented on tests/baseline/test_effective_ripple_1D.png:
any idea why this changed so much?

f0uriest commented on tests/test_interp_utils.py:
```diff
@@ -396,6 +421,99 @@ def test_fourier_chebyshev(self, func, m, n):
+    @pytest.mark.mpl_image_compare(remove_text=True, tolerance=tol_2d)
+    @staticmethod
+    def test_lambda_fourier_vartheta_zeta(name):
+        """Plot Fourier spectrum of Λ(ϑ, ζ)."""
```
is this lowercase lambda? (ie our usual poloidal stream function?)

unalmis replied:
yes Λ is DESC stream. little lambda is pitch angle

f0uriest commented on desc/integrals/_bounce_utils.py:
```diff
@@ -617,84 +638,84 @@ def fast_chebyshev(T, f, Y, num_theta, m_modes, n_modes, NFP=1, *, vander=None):
-    num_theta,
-    m_modes,
-    n_modes,
+    num_θ,
```
still not resolved

f0uriest commented on desc/integrals/_bounce_utils.py:
```diff
@@ -705,62 +731,72 @@ def fast_cubic_spline(
 
-    zeta = jnp.ravel(zeta + (T.domain[1] - T.domain[0]) * jnp.arange(T.X)[:, None])
-    f = CubicSpline(x=zeta, y=f, axis=-1, check=check).c
+    ζ = jnp.ravel(
```
unicode

f0uriest commented on desc/integrals/_bounce_utils.py:
```diff
@@ -769,46 +805,123 @@ def _move(f, out=True):
+
+def Y_B_rule(Y, NFP, spline=True):
+    """Guess Y_B from resolution of Chebyshev spectrum of angle."""
+    return (2 * Y * int(np.sqrt(NFP))) if spline else Y
```
where does the `int(sqrt(NFP))` come from?

f0uriest commented on desc/integrals/basis.py:
```diff
@@ -256,14 +257,187 @@ def compute_cheb(self, x):
+        )
+        BUG = jax.__version_info__ <= (0, 7, 2)
+        errorif(
+            self._c.ndim > 2 and BUG,
```
this would only be an issue if we're approximating a vector valued qty correct? ie `self._c.shape == (..., X, Y)`?

f0uriest commented on desc/integrals/bounce_integral.py:
```diff
@@ -422,121 +622,154 @@ def compute_theta(
         **kwargs,
     ):
-        """Return DESC coordinates θ of (α,ζ) Fourier Chebyshev basis nodes.
+        """Return the angle for mapping boundary coordinates to field line coordinates.
```
what does "boundary coordinates" refer to here?

unalmis replied:
Boundary coordinates are DESC theta and zeta always. I figured boundary is easier to swallow than understanding our symbols.

f0uriest commented on desc/integrals/bounce_integral.py:
```diff
@@ -760,11 +996,12 @@ def _nufft(self, zeta, theta, data, eps):
-        data = {name: _mmt_for_bounce(v, c) for name, c in data.items()}
-        data["B^zeta"] = _mmt_for_bounce(v, self._c["B^zeta"])
-        data["|B|"] = _mmt_for_bounce(v, self._c["|B|"])
+    def _nummt(self, ζ, data):
```
unicode

f0uriest commented on desc/integrals/bounce_integral.py:
```diff
@@ -812,99 +1049,98 @@ def interp_to_argmin(self, f, points, *, nufft_eps=1e-6, is_fourier=False):
             sentinel=0.0,
         )
-        theta = self._c["T(z)"].eval1d(ext)
+        θ = self._theta.eval1d(ext)
```
unicode

f0uriest commented on desc/integrals/bounce_integral.py:
```diff
@@ -812,99 +1049,98 @@ def interp_to_argmin(self, f, points, *, nufft_eps=1e-6, is_fourier=False):
+        -----
+        This result will converge to
+        (num transit / 2π) * ∬_Ω abs(𝐁⋅∇ζ)⁻¹ dα dζ where (α,ζ) ∈ Ω = [0, 2π)².
+        In new versions of DESC, this can be computed more efficiently as
```
it looks like its not currently tested?

f0uriest commented on desc/integrals/bounce_integral.py:
```diff
@@ -984,36 +1220,119 @@ def plot_theta(self, l, m, **kwargs):
+        kwargs
+            Keyword arguments to pass to ``matplotlib``.
 
-    The bounce integral is defined as ∫ f(ρ,α,λ,ℓ) dℓ where
```
see note above, we may want to keep this for RTD

f0uriest commented on desc/integrals/bounce_integral.py:
```diff
@@ -1332,24 +1723,22 @@ def _integrate(
-        zeta = flatten_mat(
-            bijection_from_disc(x, z1[..., None], z2[..., None]), 2 + batch
-        )
+        ζ = flatten_mat(bijection_from_disc(x, z1[..., None], z2[..., None]), 2 + batch)
```
unicode

f0uriest commented on desc/compute/_neoclassical.py:
```diff
@@ -86,70 +76,41 @@
+def _field_line_weight(params, transforms, profiles, data, **kwargs):
+    """∬_Ω abs(𝐁⋅∇ζ)⁻¹ dα dζ where (α,ζ) ∈ Ω = [0, 2π)².
 
+    The returned quantity has shape (num rho, ).
```
I'd prefer if this was expanded out to match with the other qtys, I don't see any strong reason to make an exception just for this.

f0uriest commented on desc/compute/_neoclassical.py:
```diff
@@ -86,70 +76,41 @@
-        grid.compress(data["max_tz |B|"]),
-        num_pitch,
-        simp=simp,
+    data["V_r/psi_r"] = surface_integrals(
```
We already have compute qtys for `V_r(r)` and `psi_r`, isn't this just the same as `data['V_r(r)'] / data['psi_r'])`

unalmis replied:
anything that differs from the implementation i have written makes it a total pain to do the axis limit

f0uriest commented on tests/test_axis_limits.py:
```diff
@@ -252,7 +252,9 @@ def assert_is_continuous(
             continue
         # make single variable function of rho
-        if data_index[p][name]["coordinates"] == "r":
+        if name == "V_r/psi_r":
```
why the special case here? can't we just add to not implemented limits?

unalmis commented on desc/equilibrium/coords.py:
```diff
@@ -500,40 +509,70 @@ def _map_clebsch_coordinates(
-        return 1 + dc_dt
+    def jacfun(t, varepsilon, q_m):
+        dq_dt = ((1j * jnp.exp(1j * modes * t) * q_m).real * modes).sum()
+        return 1 + dq_dt
```
The left hand side in this image is a function of theta. The right hand side is varepsilon, which is always a constant. We solve f(theta) = constant.

<img width="884" height="168" alt="image" src="https://github.com/user-attachments/assets/013ef2f6-c988-4463-8057-5720693f2e64" />

varepsilon is also always continuous along field lines, regardless of where you define the start and end of the field line, so that's why i chose it as the constant rather than solving something else like: f(theta) - iota zeta = varepsilon - iota zeta.

unalmis commented on desc/integrals/_bounce_utils.py:
```diff
@@ -467,148 +481,160 @@ def argmin(z1, z2, f, ext, g_ext):
+        Likewise, a Chebyshev series interpolates δ(α, ζ) at fixed α.
+
+    [3] The angle θ must be continuous at the branch cuts.
+        θ = αᵢ + δ(α=αᵢ, ζ=k 2π/NFP) = αᵢ₊₁ + δ(α=αᵢ₊₁, ζ=0) for k ∈ ℤ.
```
I believe k is right. I wrote a simpler explanation in the paper, so I intend to removed this comment in the code.

<img width="922" height="186" alt="image" src="https://github.com/user-attachments/assets/38d7415d-d47c-4aca-8982-aa6f4bd156a7" />
Define the left hand side as f. We are solving f(theta) = constant.

where `constant` = alpha + iota zeta.

Now I solve this equation on a single field period, and want to know what the solution to the equation is at some zeta outside this field period. If I mod zeta by 2pi/NFP, then f(theta, zeta) doesn't change since it's NFP periodic. So you just have to shift alpha by some amount to ensure that `constant` stays the same.

<img width="922" height="303" alt="image" src="https://github.com/user-attachments/assets/dfbb2416-a8f0-4c3c-895d-ee452165d15e" />

unalmis commented on desc/integrals/_bounce_utils.py:
```diff
@@ -467,148 +481,160 @@ def argmin(z1, z2, f, ext, g_ext):
     )
-    return T
+    alpha = alpha.swapaxes(0, -2)
+    delta = delta.at[..., 0].add(alpha)
```
correct, once i add alpha it becomes theta.

unalmis commented on desc/integrals/_bounce_utils.py:
```diff
@@ -769,46 +805,123 @@ def _move(f, out=True):
+
+def truncate_rule(Y):
+    """Truncation of Chebyshev series to reduce spectral aliasing."""
+    return max(1, 7 * Y // 8)
```
Just to ensure the readers are on the same page, note that sometimes one should consider an interpolating approximator as closer to the truth than a projection approximator. That is not the case here; here the truth is better modeled by the projection.

1. I did a parameter scan across various resolutions. For the range of resolutions that make any sense to use, e.g. from Y = 8 to 64, the index of the mode at which the error of the interpolant (green) approached the magnitude of its spectral coefficients (orange) was typically around `7/8 Y`. At such a point, those spectral coefficients are junk.


<img width="608" height="378" alt="image" src="https://github.com/user-attachments/assets/d62200ab-7da0-4afe-ae31-8d22957286e8" />

2. In this application we greatly prefer the galerkin approximator to the pseudospectral one because we are evaluating the function at so many points. Since it's really expensive to keep more modes, we prefer if all the error is due to truncation (galerkin) instead of both error from interpolation + truncation. It would actually be better to have a more aggressive rule since the evaluation grid is much denser, but the NFP compression made the convergence fast enough that I think it's fine.

3. i'd argue you always need some form of truncation with the chebyshev interpolation because the points cluster, and so the higher order modes you have, the more likely they are sensitive to noise from errors in root finding. removing some of the highest order ones guards against this.

unalmis replied:
> shouldn't this depend in some way on the resolution/bandwidth of the underlying function being approximated relative to the resolution used for the approximation?

You choose that dependence when you select `Y`. The assumption is you made a good choice for `Y`. Given that good choice, a truncation is preferred for reasons discussed above.

unalmis commented on desc/integrals/bounce_integral.py:
```diff
@@ -308,62 +279,291 @@ def __init__(
+            bounce integrals assuming ``fun_data`` holds all required quantities
+            to construct a ``Bounce2D`` operator as well as call its methods with
+            the flag ``is_fourier=True``.
+        fun_data : dict[str, jnp.ndarray]
```
We interpolate a Fourier series to everything in fun_data. That's typically about 3 smooth quantities that we use to compose a singular integrand with primitive functions.  In contrast, DESC data has a thousand dependencies stored in it.

unalmis replied:
see effective ripple 3/2 compute function for example

unalmis committed:
Rebasing #1919 to clean up commit history from past squash PRs

unalmis committed:
Pull arccos out of checkpoint

unalmis committed:
Add the default kwargs logic to plot_gammac

unalmis committed:
Edge case in poly_root and enable cse

unalmis committed:
Merge branch 'master' into ku/bounce

unalmis committed:
Merge branch 'master' into ku/bounce

unalmis committed:
Merge branch 'master' into ku/bounce

unalmis committed:
Merge branch 'master' into ku/bounce

unalmis committed:
improve convergence in interp_to_argmin

unalmis committed:
add a comment as requested by reviewers

unalmis committed:
improve comment

unalmis committed:
review dog being weird

YigitElma reviewed (commented):
I assume there is no math error, I couldn't check it very well because it is really hard to follow the code with all the different `angle` calls and some of function names are ambigious.

Are you planning to maintain the `interpax_fft` package?

unalmis committed:
Resolving @YigitElma comments. and replying to comment: yes i am a maintainer interpax_fft.

YigitElma commented on desc/external/paraview.py:
```diff
@@ -7,11 +7,13 @@
-        "package which is an optional dependency. Please pip install "
-        "`pyvista` to your environment."
-    )
+    with warnings.catch_warnings():
```
I am not criticizing your change. Currently, if you import `NeoIO` you get the warning for `pyvista`, this is weird for a user. If we don't have anything in init, and only allow imports like `from desc.external.neo import NeoIO` or `from desc.external.paraview import export_surface_to_paraview` etc, we won't have this problem. In future, we can have multiple optional dependencies just for external codes to work, and all of them will throw a warning

YigitElma commented on CHANGELOG.md:
```diff
@@ -26,10 +32,14 @@ Bug Fixes
 
 - No longer uses the full Hessian to compute the scale when ``x_scale="auto"`` and using a scipy optimizer that approximates the hessian (e.g. if using ``"scipy-bfgs"``, no longer attempts the Hessian computation to get the x_scale).
 - ``SplineMagneticField.from_field()`` correctly uses the ``NFP`` input when given. Also adds this as a similar input option to ``MagneticField.save_mgrid()``.
+- Significantly improves convergence of inverse stream maps in bounce integrals.
```
duplicate

unalmis committed:
Merge branch 'master' into ku/bounce

unalmis committed:
because people wanted #2059 ..

unalmis committed:
Merge branch 'master' into ku/bounce

unalmis committed:
Merge branch 'master' into ku/bounce

unalmis committed:
Merge branch 'master' into ku/bounce

unalmis committed:
Merge branch 'master' into ku/bounce

unalmis committed:
Fix autodiff at bifurcation (#2072)

When drafting a reply to a reviewer comment, I realized that the atomic
derivative computed by the autodiff tool for the `spline=True` option is
not correct if the bounce point lies near a local maxima. The
`spline=False` option is fine. It is unlikely this would have affected
optimization. 

See section 3 of 

[autodiff.pdf](https://github.com/user-attachments/files/24988182/autodiff.pdf)

unalmis committed:
Merge branch 'master' into ku/bounce

unalmis committed:
Merge branch 'master' into ku/bounce

unalmis committed:
Merge branch 'master' into ku/bounce

unalmis committed:
Merge branch 'master' into ku/bounce

unalmis committed:
Merge branch 'master' into ku/bounce

unalmis committed:
Merge branch 'master' into ku/bounce

unalmis committed:
Merge branch 'master' into ku/bounce

unalmis committed:
Merge branch 'master' into ku/bounce

unalmis committed:
merge fixs

unalmis committed:
Merge branch 'master' into ku/bounce

unalmis committed:
fix failure due to #2079 or #2065

unalmis committed:
Merge branch 'master' into ku/bounce

dpanici committed:
Merge branch 'master' into ku/bounce

YigitElma commented:
> * To now state that you are blocking the work because of that dependency is unreasonable

I am not blocking it. If others approve it, you can merge it; that is fine. I don't have to approve all PRs. I've reviewed it multiple times and spent a couple of days on it.

unalmis committed:
Merge branch 'master' into ku/bounce

unalmis committed:
Update jax-finufft version constraint in requirements

dpanici committed:
Merge branch 'master' into ku/bounce

dpanici committed:
reduce eps further for gamma_C tests

unalmis committed:
Merge branch 'master' into ku/bounce

unalmis commented on desc/integrals/_interp_utils.py:
```diff
@@ -819,20 +355,47 @@ def polyroot_vec(
+    dr = jnp.where(
+        r == sentinel,
+        0.0,
+        (jnp.expand_dims(dk, -1) - poly_val(x=r, c=dc[..., None, :])) / dc_dr,
```
Same as https://github.com/PlasmaControl/DESC/pull/2145#discussion_r3018605864.

unalmis committed:
Support dynamic shape and newton step (#2145)

- [x] resolves #1303 .
- [x] Activates newton step to find bounce points.
- It can be shown there exists O(sqrt(epsilon)) error for bounce
integrals with 1/v_ll where epsilon is error of bounce point. For v_||
integrals error is conveniently O(epsilon times sqrt(epsilon)). Hence,
the `spline` method would require thousands of knots per transit for
just a couple digits of accuracy, and it would stop convergence at
epsilon<=1e-5 (so sqrt(epsilon) <=3 digits) error due to condition
number. Of course, the `spline=False` method has always computed the
points with spectral accuracy and has very fast convergence after #1919
; that method converges to epsilon = machine precision without the
newton step. With the newton step, fast convergence is achieved with the
spline method as well.
  - I suspect fast ion confinement optimization will be easier now.

I did this on a couple lunch breaks, so it would be very weird if
clicking the approve button to merge into `master` took a year.

## Benchmarks
Here is a timing benchmark on my CPU with `nufft_eps=1e-6`. Prior to
this PR, every adjoint call to nufft1 took `>= 1 second` and the full
computation was `34 seconds`. Now every adjoint call to nufft1 is `250
milliseconds`, and the full computation is `14 seconds`. These
`improvements become larger` as the `sparsity` grows and error tolerance
parameter for the nuffts epsilon tends to 0. Likewise, the _improvement_
grows linearly with the `problem size`. As this is called within an
optimization loop where time and memory are tight, the improvement is
significant.

### Before
<img width="510" height="135" alt="Screenshot From 2026-03-29 15-22-56"
src="https://github.com/user-attachments/assets/0baf9a88-b775-4a9c-bb39-db568099aae7"
/>
<img width="510" height="135" alt="Screenshot From 2026-03-29 15-22-36"
src="https://github.com/user-attachments/assets/2bc1f730-56af-43ad-9c5e-45e9991c0bd1"
/>

### After
<img width="510" height="135" alt="Screenshot From 2026-03-29 15-12-35"
src="https://github.com/user-attachments/assets/306a6c34-32a3-46e3-8b8c-c4f259a32169"
/>
<img width="510" height="135" alt="Screenshot From 2026-03-29 15-15-59"
src="https://github.com/user-attachments/assets/db12ecb3-210a-4732-a5fc-666ba47475d8"
/>

unalmis committed:
Merge branch 'master' into ku/bounce

unalmis committed:
remove uncalled code

unalmis committed:
.

unalmis committed:
clean up supplment. final commit

unalmis committed:
pass test 4 new jax; update paper link in docs

unalmis committed:
appease whatever frakenbuild the github ci failed on
